packcheck.sh

```diff
@@ -1013,47 +1013,97 @@ ensure_cabal() {
       die "Cabal version should at least be $MIN_CABALVER"
 }
 
+# Find a file in the parent/ancestor directories
+find_in_parents () {
+  local curdir="$1" # must be absolute path without ..
+  local file="$2"
+
+  if test -e "$curdir/$file"
+  then echo "$curdir/$file"
+  elif test "$curdir" != "/"
+  then find_in_parents "$(dirname "$curdir")" "$file"
+  fi
+}
+
 ensure_cabal_project() {
   CABALCMD=cabal
   SDIST_CABALCMD=$CABALCMD
 
   if test -n "$CABAL_PROJECT"
   then
-    require_file $CABAL_PROJECT
+    require_file "$CABAL_PROJECT"
     SDIST_CABALCMD="$CABALCMD --project-file=$CABAL_PROJECT"
     if test -n "$DISABLE_SDIST_BUILD"
-    then
+    then # regular build
       CABALCMD="$SDIST_CABALCMD"
-    else
+    else # sdist build
       # We run the command from .packcheck/<package> dir after unpacking
       # sdist
       CABALCMD="$CABALCMD --project-file=../../$CABAL_PROJECT"
     fi
   fi
-  echo "Using cabal command [$CABALCMD]"
-}
 
-ensure_stack_yaml() {
-  if test -n "$STACK_YAML"
-  then
-    require_file $STACK_YAML
+  if test -n "$CABAL_PROJECT"
+  then # using explicit cabal project file
+    echo "Using user specified cabal project file at [$CABAL_PROJECT]"
+    if test -z "$DISABLE_SDIST_BUILD"
+    then # using sdist build
+      # XXX should return an error instead?
+      echo "WARNING!! You should not test a distribution build with a cabal" \
+           "project file. It may not be reproducible"
+    fi
   else
-    require_file stack.yaml
+    local implicit_proj_file
+    implicit_proj_file=$(find_in_parents "$(pwd)" cabal.project)
+
+    if test -n "$implicit_proj_file"
+    then
+      echo "Implicit cabal project file found at [$implicit_proj_file]"
+    fi
   fi
+  echo "Using cabal command [$CABALCMD]"
+}
 
+# XXX update this too
+ensure_stack_yaml() {
   SDIST_STACKCMD=$STACKCMD
   if test -n "$STACK_YAML"
   then
+    require_file $STACK_YAML
     SDIST_STACKCMD="$STACKCMD --stack-yaml $STACK_YAML"
     if test -n "$DISABLE_SDIST_BUILD"
-    then
+    then # regular build
       STACKCMD="$SDIST_STACKCMD"
-    else
+    else # sdist build
       # We run the stack command from .packcheck/<package> dir after unpacking
       # sdist
       STACKCMD="$STACKCMD --stack-yaml ../../$STACK_YAML"
     fi
   fi
+
+  if test -n "$STACK_YAML"
```

Comment:
Can this be merged with the above conditional. They seem to have a lot of similar branches.

Reply:
Merged.